File: widget/src/Features.swift
import Foundation

// THE FEATURE REGISTRY, app side. The full design is the header of
// connectors/lib/features.mjs and the short version is ops/FEATURES.md; this
// file is the same contract for Swift and must not drift from either.
//
// One sentence of it, because it is the part that bites: a missing or
// unreadable registry resolves to `allOff`, never to "everything on". The app
// ships the file inside its own bundle (widget/build.sh copies ops/features.json
// to backend/ops/features.json), so an unreadable one means the bundle is
// broken, and a broken bundle must not provision a Matrix homeserver.
//
// The owner override, ~/.hazlie/features.json, is partial and merged over the
// shipped file. An unknown key there is logged and the whole override is
// DISCARDED — the shipped registry still applies. A typo in a developer's local
// file must not brick the app.

/// A connector's three states. `optional` is not a boolean in disguise: it
/// means offered on the connections page but not auto-scheduled until the owner
/// connects it.
enum ConnectorState: Equatable {
  case on
  case off
  case optional

  init(_ value: Bool) { self = value ? .on : .off }
}

/// WHY the answer is `allOff`, when it is. The same three words node's loader
/// uses (connectors/lib/features.mjs), because the shelf and the daemon both
/// have to say the same thing about the same broken install.
///
/// A missing or invalid registry switches off the card's own connectors too —
/// imessage, mail, calendar, contacts — which is a total product outage, and
/// until this existed it was indistinguishable at every surface from an install
/// where the owner had simply connected nothing.
enum RegistryState: String {
  case ok
  case missing
  case invalid
}

struct FeatureSet: Equatable {
  var chat = false
  var voice = false
  var bridges = false
  var timeline = false
  var constellation = false
  var distiller = false
  var frontierHandoff = false
  var search = false
  var connectors: [String: ConnectorState] = [:]

  /// The order features are logged in; also the closed set of known names.
  static let names = ["chat", "voice", "bridges", "timeline", "constellation",
                      "distiller", "frontierHandoff", "search"]
  static let connectorNames = ["imessage", "mail", "calendar", "contacts", "linkedin",
                               "whatsapp", "granola",
                               "notes", "files", "photos", "notion", "oura", "health"]

  /// Everything off, including the card's own connectors. The static default
  /// and the answer to any unreadable registry.
  static let allOff = FeatureSet(
    connectors: Dictionary(uniqueKeysWithValues: connectorNames.map { ($0, ConnectorState.off) }))

  func connector(_ name: String) -> ConnectorState { connectors[name] ?? .off }

  subscript(name: String) -> Bool {
    switch name {
    case "chat": return chat
    case "voice": return voice
    case "bridges": return bridges
    case "timeline": return timeline
    case "constellation": return constellation
    case "distiller": return distiller
    case "frontierHandoff": return frontierHandoff
    case "search": return search
    default: return false
    }
  }

  /// Names only. Never the file, never anything a reader could mistake for
  /// owner data — the same rule the node logger holds to.
  var enabledNames: [String] {
    FeatureSet.names.filter { self[$0] }
      + FeatureSet.connectorNames.filter { connector($0) == .on }.map { "connector:\($0)" }
      + FeatureSet.connectorNames.filter { connector($0) == .optional }.map { "connector:\($0)?" }
  }
}

enum Features {
  private static let lock = NSLock()
  private static var cached: FeatureSet?
  private static var cachedState: RegistryState?

  /// The effective set for this process. Read once; the registry ships in the
  /// bundle and the override is a developer affordance, so re-reading it on
  /// every decision would buy nothing and would put a file read on the path of
  /// every panel open.
  static var current: FeatureSet {
    lock.lock()
    defer { lock.unlock() }
    if let cached { return cached }
    let (value, state) = loadWithState()
    cached = value
    cachedState = state
    return value
  }

  /// 'ok', 'missing' or 'invalid' for the set `current` returned. Read through
  /// the same cache, so asking costs nothing and cannot disagree with it.
  static var registryState: RegistryState {
    _ = current
    lock.lock()
    defer { lock.unlock() }
    return cachedState ?? .invalid
  }

  static func on(_ name: String) -> Bool { current[name] }
  static func connector(_ name: String) -> ConnectorState { current.connector(name) }

  /// Log what is ON, once, at launch. A build that turns something back on
  /// should say so in the same place a reader looks for why it did.
  static func logEnabled() {
    let names = current.enabledNames
    NSLog("Intaglio Labs: features on — \(names.isEmpty ? "(none)" : names.joined(separator: ", "))")
  }

  // MARK: decisions
  //
  // Pure functions of a FeatureSet, so the provisioning decisions can be read
  // (and tested) without standing up a Mac. Each one answers a question some
  // caller in Provision/main was previously answering by not asking.

  /// Warm the Matrix/mautrix runtime at launch? ~305 MB of binaries and a
  /// Synapse build, for sources the card does not read.
  static func shouldPrefetchBridgeRuntime(_ f: FeatureSet) -> Bool { f.bridges }

  /// Run setup-bridges-native.sh, which is what installs io.intaglio.bridges.
  static func shouldEnsureBridgeRuntime(_ f: FeatureSet) -> Bool { f.bridges }

  /// Retire an io.intaglio.bridges agent a previous install left running.
  /// Skipping the install is not enough on an upgrade: launchd already has the
  /// job, and RunAtLoad + KeepAlive means it comes back every login.
  static func shouldRetireBridgesAgent(_ f: FeatureSet) -> Bool { !f.bridges }

  /// Clone the ~495 MB voice model tree out of the bundle into ~/.hazlie.
  /// (Taking the models OUT of the bundle is stage 2; this only stops the copy.)
  static func shouldCloneVoiceModels(_ f: FeatureSet) -> Bool { f.voice }

  /// Build the hidden ear webview inside the widget window.
  static func shouldBuildEarWebView(_ f: FeatureSet) -> Bool { f.voice }

  /// Build the chat panel and its webview.
  static func shouldBuildChatPanel(_ f: FeatureSet) -> Bool { f.chat }

  /// Build the timeline (people-months) panel.
  static func shouldBuildTimelinePanel(_ f: FeatureSet) -> Bool { f.timeline }

  /// ~~`peopleButtonOpensPeopleDirectly`: with the timeline off, route the
  /// People button straight to the "Same person?" review, which was a KEEP and
  /// had no other door.~~ Gone with the surface review (2026-09-13): the review
  /// IS the duplicate-pair window the owner asked to take off the bar, so the
  /// button follows `timeline` in both halves now — hidden in widget.js before
  /// the first paint, and `openMonths()` returning early in main.swift. The
  /// pages and their bridge grants stay, unreachable rather than deleted; a
  /// later `identity` flag is what would separate find-pairs from the timeline.
  /// `shouldBuildTimelinePanel` is the one gate both halves read.

  /// Run a distillation pass. ANDed with the existing ~/.hazlie/distill.enabled
  /// marker, never replacing it: the marker is the "no in-app way to review
  /// claims yet" gate and it has its own reason to exist.
  static func shouldDistill(_ f: FeatureSet, markerPresent: Bool) -> Bool {
    f.distiller && markerPresent
  }

  // MARK: loading

  private static var bundledRegistry: URL? {
    (Bundle.main.resourceURL ?? Bundle.main.bundleURL)
      .appendingPathComponent("backend/ops/features.json")
  }

  private static var ownerOverride: URL {
    FileManager.default.homeDirectoryForCurrentUser
      .appendingPathComponent(".hazlie/features.json")
  }

  static func load(registry: URL? = bundledRegistry, override: URL? = ownerOverride) -> FeatureSet {
    loadWithState(registry: registry, override: override).0
  }

  /// The set AND why. `missing` is a bundle that does not carry the file at
  /// all; `invalid` is one that carries an unreadable one. Both answer allOff —
  /// "the file that says what is on is unreadable" must never resolve to
  /// "everything is on" — and the owner is owed different words for each.
  static func loadWithState(registry: URL? = bundledRegistry,
                            override: URL? = ownerOverride) -> (FeatureSet, RegistryState) {
    guard let registry, let data = try? Data(contentsOf: registry) else {
      NSLog("Intaglio Labs: no feature registry in this bundle — everything is off")
      return (.allOff, .missing)
    }
    guard let base = parseRegistry(data) else {
      NSLog("Intaglio Labs: the feature registry is unreadable — everything is off")
      return (.allOff, .invalid)
    }
    guard let override, let overrideData = try? Data(contentsOf: override) else { return (base, .ok) }
    guard let object = try? JSONSerialization.jsonObject(with: overrideData) as? [String: Any] else {
      NSLog("Intaglio Labs: features override is not a JSON object — ignored")
      return (base, .ok)
    }
    do {
      return (try merge(base, object), .ok)
    } catch {
      // DISCARD THE OVERRIDE, KEEP THE REGISTRY. See the header.
      NSLog("Intaglio Labs: features override ignored: \(error.localizedDescription)")
      return (base, .ok)
    }
  }

  static func parseRegistry(_ data: Data) -> FeatureSet? {
    guard let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
          jsonNumber(object["version"])?.intValue == 1,
          let features = object["features"] as? [String: Any] else { return nil }
    // Start from allOff, so a key the shipped file forgot is OFF rather than
    // inheriting whatever a struct default happened to be.
    return try? merge(.allOff, features)
  }

  /// A JSON `true`/`false`, AND NOT A NUMBER THAT LOOKS LIKE ONE.
  ///
  /// JSONSerialization hands back NSNumber for both `true` and `1`, and
  /// `as? Bool` accepts either — so `{"chat": 1}` switched chat ON in the app
  /// while node's loader (which requires `typeof value === 'boolean'`) threw the
  /// WHOLE override away and reported chat off. One file, two answers, and the
  /// daemon and the app acting on different registries. CFBooleanGetTypeID is
  /// the one check that tells the two apart: JSON booleans arrive as
  /// kCFBoolean{True,False}, JSON numbers as __NSCFNumber.
  static func jsonBool(_ value: Any) -> Bool? {
    guard CFGetTypeID(value as CFTypeRef) == CFBooleanGetTypeID() else { return nil }
    return (value as? NSNumber)?.boolValue
  }

  /// A JSON NUMBER, AND NOT A BOOLEAN WEARING ONE'S CLOTHES. The same NSNumber
  /// ambiguity as jsonBool, one field further up and pointing the other way:
  /// `as? Int` on the NSNumber that JSON `true` bridges to yields 1, so
  /// `{"version": true}` passed this app's version check while node's loader
  /// (`raw?.version !== 1`) refused the whole file. A registry shipped that way
  /// meant the daemon at ALL_OFF, scheduling nothing and painting the shelf's
  /// red line, while the app ran the full feature set — chat panel and all. Two
  /// processes, one file, opposite answers, which is the exact failure jsonBool
  /// exists to prevent one field below.
  static func jsonNumber(_ value: Any?) -> NSNumber? {
    guard let value, CFGetTypeID(value as CFTypeRef) != CFBooleanGetTypeID() else { return nil }
    return value as? NSNumber
  }

  private struct Rejected: LocalizedError {
    let reason: String
    var errorDescription: String? { reason }
  }

  static func merge(_ base: FeatureSet, _ object: [String: Any]) throws -> FeatureSet {
    var merged = base
    for (key, value) in object {
      if key == "version" { continue } // informational; an override is partial by design
      if key == "connectors" {
        guard let table = value as? [String: Any] else {
          throw Rejected(reason: "\"connectors\" must be an object")
        }
        for (name, state) in table {
          guard FeatureSet.connectorNames.contains(name) else {
            throw Rejected(reason: "unknown connector \"\(name)\"")
          }
          // NSNumber bridges JSON true/false, so check the string case first and
          // the boolean second; `as? Bool` on the string would simply fail, but
          // saying the order out loud is cheaper than rediscovering it.
          if let text = state as? String {
            guard text == "optional" else {
              throw Rejected(reason: "connector \"\(name)\" must be true, false or \"optional\"")
            }
            merged.connectors[name] = .optional
          } else if let flag = jsonBool(state) {
            merged.connectors[name] = ConnectorState(flag)
          } else {
            throw Rejected(reason: "connector \"\(name)\" must be true, false or \"optional\"")
          }
        }
        continue
      }
      guard FeatureSet.names.contains(key) else {
        throw Rejected(reason: "unknown feature \"\(key)\"")
      }
      guard let flag = jsonBool(value) else {
        throw Rejected(reason: "feature \"\(key)\" must be true or false")
      }
      switch key {
      case "chat": merged.chat = flag
      case "voice": merged.voice = flag
      case "bridges": merged.bridges = flag
      case "timeline": merged.timeline = flag
      case "constellation": merged.constellation = flag
      case "distiller": merged.distiller = flag
      case "frontierHandoff": merged.frontierHandoff = flag
      case "search": merged.search = flag
      default: break // unreachable: names is checked above
      }
    }
    return merged
  }
}

File: connectors/lib/features.mjs
// THE FEATURE REGISTRY — one file, read by everything.
//
// `ops/features.json` says which of this app's surfaces are alive. Stage 1 of
// the "Reconnect, Only" repackaging turns everything the reconnection card
// does not need OFF, without deleting a line of it: the code stays in the
// repository, the feature is behind a flag that defaults off, its provisioning
// step is skipped, and its launch agent is not installed.
//
// WHY A FILE AND NOT A CONSTANT. Three processes have to agree — the Swift app,
// hermes, and the connectors daemon — and they are built and shipped together
// but started apart. A constant in one of them is a fourth opinion. The file is
// copied into the bundle by widget/build.sh (to backend/ops/features.json), so
// the repo layout and the bundle layout resolve it at the SAME relative path
// from here: ../../ops/features.json is repo/ops in a checkout and $BE/ops in
// the app. Do not "tidy" that into an absolute path.
//
// OWNER OVERRIDE. ~/.hazlie/features.json, same shape, partial allowed, merged
// over the shipped file. That is what lets a developer switch chat back on for
// an afternoon without a rebuild, and it is also the escape hatch if a flag
// turns out to have been wrong on a machine that cannot be rebuilt today.
//
// FAIL CLOSED, EXCEPT ON THE OVERRIDE. A missing or malformed registry yields
// ALL_OFF — the card's own connectors included — because "the file that says
// what is on is unreadable" must never resolve to "everything is on". The
// override is the deliberate exception: an unknown key or a bad value there is
// REPORTED and the override is DISCARDED, and the shipped registry still
// applies. A fatal override is how ~/.hazlie/connectors/config.json's closed-key
// assertion once killed the connector daemon on a single typo; this file will
// not repeat that.
//
// CONNECTOR VALUES ARE THREE-STATE, and the third one is not a boolean in
// disguise:
//   true       — the card reads it. Scheduled, tile shown.
//   false      — dormant. Disabled-marker semantics in the daemon (never
//                scheduled, logged as source_hidden), tile hidden.
//   'optional' — a real participant source, small today. Offered on the
//                connections page and labelled as optional, but NOT
//                auto-scheduled until the owner connects it; its own gate
//                (WhatsApp's .disabled marker, Granola's credential check via
//                needs()) is what lets it start. This is exactly what WhatsApp
//                already did, now written down instead of implied.
//
// `health` names no connector module today — Apple Health was replaced by the
// Oura API in 2026-08. It is listed anyway so that re-adding it cannot arrive
// switched on by inheriting a missing key.

import { readFileSync } from 'node:fs';
import { homedir } from 'node:os';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

const here = dirname(fileURLToPath(import.meta.url));

/// Same relative path in a checkout and inside the app bundle. See the header.
export const DEFAULT_REGISTRY_PATH = join(here, '..', '..', 'ops', 'features.json');

/// The owner override's path — and the one knob a TEST is allowed to turn.
///
/// HAZLIE_FEATURES_OVERRIDE names another file, or the literal 'none' for "no
/// override at all". Without it, a test that pins what this install ships was
/// really reading the DEVELOPER's ~/.hazlie/features.json: the daemon reads the
/// override at module scope and hermes reads it per request, so a local
/// `{"bridges":true}` turned the policy assertions red on that machine and
/// nowhere else. The variable belongs HERE rather than in each caller, because
/// hermes, the daemon and connect all reach the registry through this function
/// and only one of them is a file a test can edit.
///
/// EMPTY IS NONE, NOT "READ $HOME". `HAZLIE_FEATURES_OVERRIDE=` is the spelling
/// a wrapper produces from a variable that was never set
/// (`HAZLIE_FEATURES_OVERRIDE=$MAYBE node --test …`), and falling through to the
/// owner file there un-hermeticized exactly the tests this knob exists to make
/// hermetic. Setting the variable at all is a decision; setting it to nothing is
/// the decision to use no override, like the literal 'none'.
export function defaultOverridePath(home = homedir(), env = process.env) {
  const configured = env?.HAZLIE_FEATURES_OVERRIDE;
  if (configured === 'none' || configured === '') return null;
  if (typeof configured === 'string') return configured;
  return join(home, '.hazlie', 'features.json');
}

/// Whether the override path was NAMED by somebody rather than being the
/// default owner file. It separates two silences: the owner file is usually
/// absent and that is the normal case, while a path a human typed that is not
/// there is a mis-pointed escape hatch — which used to be indistinguishable
/// from a clean read of a file that was never opened.
export function overrideIsConfigured(env = process.env) {
  const configured = env?.HAZLIE_FEATURES_OVERRIDE;
  return typeof configured === 'string' && configured !== '' && configured !== 'none';
}

/// The boolean surfaces. Order is the order they are logged in.
export const FEATURE_NAMES = Object.freeze([
  'chat', 'voice', 'bridges', 'timeline', 'constellation',
  'distiller', 'frontierHandoff', 'search',
]);

/// The connector surfaces. Not the same list as daemon.mjs' CONNECTOR_NAMES,
/// deliberately: `health` is here with no module, and `matrix` is a module with
/// no entry here because Matrix is the BRIDGES feature, not a source somebody
/// connects. connectorsDisabledBy() below is where the two lists meet.
export const CONNECTOR_FEATURE_NAMES = Object.freeze([
  'imessage', 'mail', 'calendar', 'contacts', 'linkedin',
  'whatsapp', 'granola',
  'notes', 'files', 'photos', 'notion', 'oura', 'health',
]);

/// The static default, and the answer whenever the registry cannot be read.
export const ALL_OFF = Object.freeze({
  ...Object.fromEntries(FEATURE_NAMES.map((name) => [name, false])),
  connectors: Object.freeze(
    Object.fromEntries(CONNECTOR_FEATURE_NAMES.map((name) => [name, false]))
  ),
});

function isConnectorValue(value) {
  return value === true || value === false || value === 'optional';
}

/// Merge a partial override over a full base. Throws — with a message naming
/// the offending key and NOTHING ELSE — on anything that is not a known key
/// with a legal value. The message carries no context prefix on purpose: this
/// is called both for the shipped registry and for the owner override, and each
/// caller says which one it was reading. Callers that must not die on a typo
/// catch it; see readFeatures.
export function mergeFeatures(base, override) {
  if (override === null || typeof override !== 'object' || Array.isArray(override)) {
    throw new Error('must be a JSON object');
  }
  const merged = { ...base, connectors: { ...base.connectors } };
  for (const [key, value] of Object.entries(override)) {
    if (key === 'version') continue; // informational; the override is partial by design
    if (key === 'connectors') {
      if (value === null || typeof value !== 'object' || Array.isArray(value)) {
        throw new Error('"connectors" must be an object');
      }
      for (const [name, state] of Object.entries(value)) {
        if (!CONNECTOR_FEATURE_NAMES.includes(name)) {
          throw new Error(`unknown connector "${name}"`);
        }
        if (!isConnectorValue(state)) {
          throw new Error(`connector "${name}" must be true, false or "optional"`);
        }
        merged.connectors[name] = state;
      }
      continue;
    }
    if (!FEATURE_NAMES.includes(key)) {
      throw new Error(`unknown feature "${key}"`);
    }
    if (typeof value !== 'boolean') {
      throw new Error(`feature "${key}" must be true or false`);
    }
    merged[key] = value;
  }
  return merged;
}

/// Parse a full registry file. Unknown keys here are the SHIPPED file's
/// problem, not a developer's, so this is strict and the caller falls back to
/// ALL_OFF rather than to a half-read registry.
export function parseRegistry(text) {
  const raw = JSON.parse(text);
  if (raw?.version !== 1) throw new Error('features registry: version must be 1');
  if (raw.features === null || typeof raw.features !== 'object' || Array.isArray(raw.features)) {
    throw new Error('features registry: "features" must be an object');
  }
  // Start from ALL_OFF so a key the shipped file forgot is OFF, not undefined.
  return mergeFeatures(ALL_OFF, raw.features);
}

/// WHY THE REGISTRY IS ALL_OFF, when it is.
///
/// 'ok'      — it was read and parsed.
/// 'missing' — the file is not there. On a shipped app that means a broken
///             bundle; in a checkout, a deleted ops/features.json.
/// 'invalid' — it is there and unreadable: bad JSON, wrong version, a key or
///             value the loader refuses.
///
/// Both of the last two resolve to ALL_OFF, which is right — "the file that
/// says what is on is unreadable" must never resolve to "everything is on" —
/// and both used to be INDISTINGUISHABLE FROM A CORRECT ALL-OFF INSTALL at
/// every surface that reads them. The connections shelf drew the same empty
/// list it draws for "nothing connected", and the only report was one line on
/// the daemon's stderr. A total product outage deserves a different sentence
/// than a quiet Tuesday, so the state travels with the set.
export const REGISTRY_STATES = Object.freeze(['ok', 'missing', 'invalid']);

/// AND WHAT BECAME OF THE OVERRIDE, which is a separate question with separate
/// consequences.
///
/// 'none'    — there was none to apply: no path, or the owner file is simply
///             not there. The normal case, and silent.
/// 'ok'      — a file was read and merged.
/// 'missing' — a path somebody NAMED is not there. The escape hatch is
///             mis-pointed, which used to look exactly like no override at all.
/// 'invalid' — it is there and was refused: bad JSON, an unknown key, an
///             illegal value. Discarded; the shipped registry still applies.
///
/// None of the last two touch `registryState`. A bad override is not a broken
/// bundle, and reporting it as one would send the owner to reinstall over a
/// typo in a file they are invited to edit.
export const OVERRIDE_STATES = Object.freeze(['none', 'ok', 'missing', 'invalid']);

/// The effective set AND why it is what it is:
/// `{ features, registryState, overrideState }`. `onProblem` is called with a
/// one-line reason for anything that had to be ignored; it never throws out of
/// here.
///
/// `overrideConfigured` says whether the path was named rather than defaulted —
/// see overrideIsConfigured. It is only ever used to decide whether an ABSENT
/// override is worth a word.
export function readFeatureRegistry({
  registryPath = DEFAULT_REGISTRY_PATH,
  overridePath = defaultOverridePath(),
  overrideConfigured = overrideIsConfigured(),
  onProblem = () => {},
} = {}) {
  let features = ALL_OFF;
  try {
    features = parseRegistry(readFileSync(registryPath, 'utf8'));
  } catch (error) {
    // ENOENT is the bundle being incomplete; anything else is the file being
    // wrong. Same ALL_OFF answer, different thing to tell the owner.
    const registryState = error?.code === 'ENOENT' ? 'missing' : 'invalid';
    onProblem(`features registry ${registryState}, everything is off: ${error.message}`);
    return { features: ALL_OFF, registryState, overrideState: 'none' };
  }
  let overrideText = null;
  try {
    // A null path is "no override", not a file at "null": HAZLIE_FEATURES_OVERRIDE=none
    // says so explicitly, and readFileSync(null) throwing into the catch below
    // would have been the right answer for the wrong reason.
    if (overridePath === null || overridePath === undefined) {
      return { features, registryState: 'ok', overrideState: 'none' };
    }
    overrideText = readFileSync(overridePath, 'utf8');
  } catch (error) {
    // THE ONE CASE THAT IS NOT NORMAL. A file that is not there is the usual
    // shape of "this machine has no override" — unless somebody named it, in
    // which case they are acting on a file nothing read. Anything other than
    // ENOENT (a directory, a permission, an unreadable device) is the file
    // existing and refusing, which is worth saying whoever chose the path.
    const overrideState = error?.code === 'ENOENT' ? 'missing' : 'invalid';
    if (overrideState === 'invalid' || overrideConfigured) {
      onProblem(`features override unreadable, ignored: ${error.message}`);
      return { features, registryState: 'ok', overrideState };
    }
    return { features, registryState: 'ok', overrideState: 'none' };
  }
  try {
    return {
      features: mergeFeatures(features, JSON.parse(overrideText)),
      registryState: 'ok',
      overrideState: 'ok',
    };
  } catch (error) {
    onProblem(`features override ignored: ${error.message}`);
    return { features, registryState: 'ok', overrideState: 'invalid' };
  }
}

/// The effective set, for the many callers that only want the answer.
export function readFeatures(options = {}) {
  return readFeatureRegistry(options).features;
}

/// Names only — never the file, never a count that would let a reader guess at
/// owner data. This is what every process logs at startup.
export function enabledFeatureNames(features) {
  return [
    ...FEATURE_NAMES.filter((name) => features[name] === true),
    ...CONNECTOR_FEATURE_NAMES
      .filter((name) => features.connectors?.[name] === true)
      .map((name) => `connector:${name}`),
    ...CONNECTOR_FEATURE_NAMES
      .filter((name) => features.connectors?.[name] === 'optional')
      .map((name) => `connector:${name}?`),
  ];
}

/// Which of the daemon's connector modules must never be scheduled.
///
/// Two rules, and the second one is the one that is easy to forget: a connector
/// whose feature is `false` is disabled, AND `matrix` is disabled whenever
/// `bridges` is off, because the Matrix bus IS the bridges feature. A module
/// with no entry in the registry at all is left alone rather than disabled —
/// silently switching off a source somebody added is worse than listing it.
export function connectorsDisabledBy(features, connectorNames) {
  return connectorNames.filter((name) => {
    if (name === 'matrix') return features.bridges !== true;
    const state = features.connectors?.[name];
    return state === false;
  });
}

/// Which connectors are offered but not auto-started.
export function optionalConnectors(features, connectorNames) {
  return connectorNames.filter((name) => features.connectors?.[name] === 'optional');
}
